libgalois/src/analytics/pagerank/pagerank-push.cpp

```diff
@@ -0,0 +1,220 @@
+using EdgeData = std::tuple<>;
+typedef katana::PropertyGraph<NodeData, EdgeData> Graph;
+typedef typename Graph::Node GNode;
+
+void
+InitializeNodeResidual(Graph& graph, katana::analytics::PagerankPlan plan) {
+  katana::do_all(
+      katana::iterate(graph),
+      [&](const GNode& n) {
+        graph.GetData<NodeResidual>(n) = plan.initial_residual();
+        graph.GetData<NodeValue>(n) = 0;
+      },
+      katana::no_stats(), katana::loopname("Initialize"));
+}
+
+}  // namespace
+
+katana::Result<void>
+PagerankPushAsync(
+    katana::PropertyFileGraph* pfg, const std::string& output_property_name,
+    katana::analytics::PagerankPlan plan) {
+  katana::Prealloc(5, 5 * pfg->topology().num_nodes() * sizeof(NodeData));
+
+  std::string temporary_property_name =
+      boost::uuids::to_string(boost::uuids::random_generator()());
+
+  if (auto result = katana::analytics::ConstructNodeProperties<NodeData>(
+          pfg, {output_property_name, temporary_property_name});
+      !result) {
+    return result.error();
+  }
+  TemporaryPropertyCleanup temporary_property_name_cleanup{
+      pfg, temporary_property_name};
+
+  auto graph_result =
+      Graph::Make(pfg, {output_property_name, temporary_property_name}, {});
+  if (!graph_result) {
+    return graph_result.error();
+  }
+  Graph graph = graph_result.value();
+
+  InitializeNodeResidual(graph, plan);
+
+  typedef katana::PerSocketChunkFIFO<
+      katana::analytics::PagerankPlan::kChunkSize>
+      WL;
+  katana::for_each(
+      katana::iterate(graph),
+      [&](const GNode& src, auto& ctx) {
+        auto& src_residual = graph.GetData<NodeResidual>(src);
+        if (src_residual > plan.tolerance()) {
+          PRTy old_residual = src_residual.exchange(0.0);
+          auto& src_value = graph.GetData<NodeValue>(src);
+          src_value += old_residual;
+          int src_nout =
+              std::distance(graph.edge_begin(src), graph.edge_end(src));
+          if (src_nout > 0) {
+            PRTy delta = old_residual * plan.alpha() / src_nout;
+            //! For each out-going neighbors.
+            for (const auto& jj : graph.edges(src)) {
+              auto dest = graph.GetEdgeDest(jj);
+              auto& dest_residual = graph.GetData<NodeResidual>(dest);
+              if (delta > 0) {
+                auto old = atomicAdd(dest_residual, delta);
+                if ((old < plan.tolerance()) &&
+                    (old + delta >= plan.tolerance())) {
+                  ctx.push(*dest);
+                }
+              }
+            }
+          }
+        }
+      },
+      katana::loopname("PushResidualAsync"),
+      katana::disable_conflict_detection(), katana::wl<WL>());
+
+  return katana::ResultSuccess();
+}
+
+katana::Result<void>
+PagerankPushSync(
+    katana::PropertyFileGraph* pfg, const std::string& output_property_name,
+    katana::analytics::PagerankPlan plan) {
+  katana::Prealloc(5, 5 * pfg->topology().num_nodes() * sizeof(NodeData));
```

Comment:
This can be simplified to use pfg->num_nodes(), so if you want to wait for your other commit to land it is up to you.

Reply:
I'll change it when the other PR drops.